[bhnuka] iP

bhnuka opened the pull request:
### Duke Chatbot

Duke is a **Personal Assistant Chatbot** that helps you keep track of various tasks. You can interact with Duke to manage your tasks and stay organized.

> "It's decent." - My mom

Manage your tasks easily with Duke:

- Add tasks
- Mark tasks as done
- ~~Delete~~ tasks
- Search for tasks

Duke can handle three task types, namely:

1. To-do tasks
2. Deadline tasks
3. Event tasks

Features (Present and future)
- [x] 3 task types
- [ ] Maybe more tasks?
- [ ] Ability to sort tasks according to due date?

Example code to interact with Duke
`duke.run();`

Ensure that the data folder is present before running

``` Java
public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
}
```

:rocket: Duke is ready to help you get organized! Just download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html) today!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

bhnuka committed:
> Gave chatbot a new name (Eva), to differentiate it from the placeholder name Duke
> Changed the ASCII Logo to Eva's
> Implemented an initial skeletal version such that it simply greets the user and exits
> Removed the initial Duke.java file

bhnuka committed:
> Improved the skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye.

bhnuka committed:
> Added the ability to store whatever text entered by the user and display them back to the user when requested.
> Assumed there will be no more than 100 tasks

bhnuka committed:
> Added the ability to mark tasks as done and to change the status back to not done.
> Used a Task class to represent tasks.

bhnuka committed:
> Added  support for tracking three types of tasks:
> ToDos: tasks without any date/time attached to it e.g., visit new theme park
> Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
> Events: tasks that start at a specific date/time and ends at a specific date/time e.g., (a) team project meeting 2/10/2019 2-4pm (b) orientation week 4/10/2019 to 11/10/2019

> Used Inheritance to support multiple task types

bhnuka committed:
> Used the input/output redirection technique to semi-automate the testing of Duke.

bhnuka committed:
> Taught the chatbot to deal with errors such as incorrect inputs entered by the user.
> Used exceptions to handle errors

bhnuka committed:
> Added support for deleting tasks from the list.
> Used ArrayList<Task> to store tasks.

bhnuka committed:
> Used Java Enums
> Added Javadoc comments where necessary

bhnuka committed:
> Cleaned up the formatting

bhnuka committed:
Add save feature and duke.txt

Tasks added will be forgotten once the program is exited

Not ideal for a list to be lost each time

Saves list to textfile before program is exited

Allows the program to remember the list upon next start up

bhnuka committed:
Add date recognition

Currently difficult to differentiate dates

Added date recognition using LocalDate and DateTimeFormatter

Libraries already available

bhnuka committed:
Refactor the code to extract out closely related code as classes

Difficult to understand as a chunk

Refactored the code

Easier to read

bhnuka committed:
Divide classes into packages

Currently unorganized

Organized into packages Eva, Task and Command

Easier to import

bhnuka committed:
Merge remote-tracking branch 'origin/add-gradle-support'

bhnuka committed:
Automate project builds using Gradle

Currently manual test

Added Gradle support

bhnuka committed:
Add JUnit tests

bhnuka committed:
Give users a way to find a task by searching for a keyword

Currently unable to search for specific tasks

Implemented finding tasks

bhnuka committed:
Tweak the code to comply with a coding standard

Should follow the standard to make it seem more cohesive

Followed the coding conventions

bhnuka committed:
Add JavaDoc comments

Lack of documentation

Added JavaDoc comments

Easiest way to write clear and concise comments

bhnuka committed:
Add a GUI to the chatbot. Use the JavaFX technology to implement the GUI.

ivyy-poison commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,82 @@
+        String[] parts = input.split(" ", 2);
+        String commandType = parts[0];
+
+        switch (commandType) {
```
I like the use of the various Command Classes to encapsulate the handling of each command

ivyy-poison commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,82 @@
+            throw new DukeException("\t OOPS!!! The description of a todo" +
+             " cannot be empty.");
+            }
+            return new AddCommand(new Todo(input.substring(5), false));
+        case "deadline":
+        // Parse deadline command and create DeadlineTask object
+            int byIndex = input.indexOf("/by");
+            if (byIndex == -1) {
+                throw new DukeException("\t OOPS!!! The deadline description" +
+                 " must include a /by date.");
+            }
+            String description = input.substring(9, byIndex).trim();
+            if (description.isEmpty()) {
+                throw new DukeException("\t OOPS!!! The deadline description" +
+                 " cannot be empty.");
+            }
+            String by = input.substring(byIndex + 3).trim();
+            return new AddCommand(new Deadline(description, false, by));
+        case "event":
+            int fromIndex = input.indexOf("/from");
+            int toIndex = input.indexOf("/to");
+            if (fromIndex == -1 && toIndex == -1) {
+                throw new DukeException("\t OOPS!!! The event description" +
+                 " must include both /from and /to dates.");
+            }
+            String description2 = input.substring(6, fromIndex).trim();
+            if (description2.isEmpty()) {
+                throw new DukeException("\t OOPS!!! The event description cannot" +
+                 " be empty.");
+            }
+            String from = input.substring(fromIndex + 5, toIndex).trim();
+            String to = input.substring(toIndex + 3).trim();
+            // Parse event command and create EventTask object
+            return new AddCommand(new Event(description2, false, from, to));
+        case "delete":
+            int taskIndex = Integer.parseInt(input.substring(7)) - 1;
+            return new DeleteCommand(taskIndex);
+        case "find":
+            String keyword = input.substring(5).trim();
+            return new FindCommand(keyword);
```
I like the use of Command Classes as a means to encapsulate the handling of various different commands!

ivyy-poison commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,82 @@
+     */
+    public static Command parse(String input) throws DukeException {
+        String[] parts = input.split(" ", 2);
+        String commandType = parts[0];
```
Perhaps consider using ENUMs instead?

ivyy-poison commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,141 @@
+    /**
+     * Returns a horizontal line separator as a String.
+     */
+    public String showLine() {
```
Perhaps store the line as a class attribute?

```suggestion
    private final String LINE = "-".repeat(60);
```

ivyy-poison commented on src/main/java/duke/task/TaskType.java:
```diff
@@ -0,0 +1,8 @@
+/**
+ * Enums to represent the different types of tasks
+ */
+public enum TaskType {
+    TODO, DEADLINE, EVENT
+}
```
I love the use of ENUMs here.

ivyy-poison reviewed (commented):
LGTM! Code is very neatly organised, and easy to read, great job :)

bhnuka committed:
changed build.gradle to compile fat jar file

bhnuka committed:
Adds assertions

Lacking assertions

Additional error handling helps prevents bugs

Added assertions

Next logical method after JUnit testing

bhnuka committed:
Improves code readability

Code has readability issues

Makes it difficult to look back on and for others to work on

Improved readability by refactoring

Easiest way to shorten methods

bhnuka committed:
Merge pull request #1 from bhnuka/branch-A-CodeQuality

Improves code readability

bhnuka committed:
Merge branch 'branch-A-Assertions'

# Conflicts:
#	data/tasks.txt
#	src/main/java/duke/Parser.java

bhnuka committed:
Adds C-DetectDuplicates

Able to add duplicates with no consequence

Makes list messy. Full of duplicate tasks.

Added functionality to AddCommand to check if added task is a duplicate

bhnuka committed:
Added User Guide and Code Refactoring

Enhanced project documentation by including a User Guide

Code was not adhering to coding standards, affecting readability and maintainability

Refactored the code to conform to established coding standards

This change ensures that anyone reviewing or contributing to the project can easily understand and follow the codebase

bhnuka committed:
Updated User Guide to Meet Project Requirements

The User Guide was incomplete and did not adhere to project documentation
standards.

Revised and expanded the User Guide to include all necessary features and
usage instructions as per project requirements.

This update ensures that users have access to comprehensive and accurate
information, improving the overall user experience.

bhnuka committed:
Updated UI Screenshot for Improved Clarity

The existing UI screenshot was outdated and did not accurately represent the current user interface of the application.

Captured an updated screenshot of the user interface to provide users with a clear and current visual representation of the application's appearance.

This update enhances the User Guide's usability and helps users better understand the application's interface.

bhnuka committed:
Refactor data file paths

Hardcoded data file paths in the code using absolute paths.

Absolute paths may not work on different computers or when creating a JAR file.

Updated code to use relative data paths instead.

To ensure the application works seamlessly when deployed on different systems.

bhnuka committed:
Update chatbot window title

The chatbot window title is set to a default value.

The window title should reflect the name of the chatbot.

Updated the UI code to set the window title to the name Duke.

To provide users with a more informative and contextually relevant window title.

bhnuka committed:
Update chatbot window title

The chatbot window title is set to a default value.

The window title should reflect the name of the chatbot.

Updated the UI code to set the window title to the name Duke.

To provide users with a more informative and contextually relevant window title.

bhnuka committed:
Enhance JUnit Tests and Implement Checkstyle

JUnit tests need improvement, and Checkstyle is not implemented

Improve code quality and test coverage for reliability

This commit enhances JUnit tests for thorough coverage and implements Checkstyle for code quality checks

Ensure code reliability and adherence to style guidelines

This update enhances code quality and test coverage
